proposals/0000-support-pun-free-code.md

```diff
@@ -0,0 +1,414 @@
+
+### Solution Overview
+
+#### Namespace qualified imports
+
+To help programmers deal with the external code that uses punning we propose
+to introduce namespace qualified import syntax, guarded behind
+`-XExplicitNamespaces` extension. The syntax introduces two markers for
+qualified imports, `data` and `type`:
+
+```haskell
+import Data.Proxy type qualified as T   -- import only the type namespace
+import Data.Proxy data qualified as D   -- import only the data namespace
+```
+
+#### Punning usage warnings
+
+We also propose to introduce two new warnings to GHC: `-Wpuns` and
+`-Wpun-bindings`.
+
+* `-Wpuns` warns the user about the usage of punning at use sites.
+
+* `-Wpun-bindings` warns the user about the introduction of punning at binding
+  sites.
+
+These warnings aim to help the user to avoid using punning in their codebase.
+
+To determine whether some binding or use site takes advantage of punning we
+ask the question: "If Haskell had a single unified namespace, would that change
+the lookup result?". If the answer is yes, then the code uses punning:
+
+```haskell
+data T = T -- with a single unified namespace, T the data would have a name clash with T the type.
+
+a = 15
+
+f :: forall a. a -> a -- no warning, because with a single unified namespace 'forall a.' would shadow the top level 'a', so 'a' in 'a -> a' would still refer to the forall bound 'a'.
+```
+
+There are more detailed examples in Examples section.
+
+## Proposed Change Specification
+
+* To disambiguate the namespaces of identifiers from modules that use punning,
+  the `-XExplicitNamespaces` extension is extended with new syntax:
+
+  ```
+  impdecl   -> import [qualified] modid [as modid] [impspec]
+  ```
+  is changed to
+  ```
+  impdecl   -> import [qualified] modid [namespace] [as modid] [impspec]
+
+  namespace -> data
+             | type
+  ```
+   * With `data` specified in the import, only identifiers belonging to the data namespace will be brought into the scope.
+   * With `type` specified in the import, only identifiers belonging to the type namespace will be brought into the scope.
+
+* Introduce a new warning, `-Wpun-bindings` and add it to `-Weverything`.
+  The warning is triggered by any name binding that would clash with another
+  identifier if Haskell had a single unified namespace.
+
+* Introduce a new warning, `-Wpuns` and add it to `-Weverything`. The warning is
+  triggered by using an identifier that would be ambiguous or refer to another
+  entity if Haskell had a single unified namespace.
+
+* Amend [GHC Proposal #65](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0065-type-infix.rst)
+  to use `data` instead of `value`.
+
+* Under `-XPatternSynonyms`, change the `pattern` qualifier in import and export lists to `data`.
+  Introduce `-Wpattern-namespace-qualifier` warning that warns when the
```

Comment:
But the previous sentence says "change", implying that e.g. `pattern True` would no longer be valid. Yet I think you mean for `pattern True` to be valid for some time still.